File: agents/iso_agent.py
from langchain_perplexity import ChatPerplexity
from typing import Dict, Any
import json
import ast
from prompts.iso_prompt import get_iso_prompt


def analyze_iso_compliance(extracted_data: Dict[str, Any], model: ChatPerplexity) -> Dict[str, Any]:
    """Analyze ISO/IEC 42001:2023 compliance"""
    
    prompt = get_iso_prompt(extracted_data)
    
    response = model.invoke(prompt)
    
    try:
        # Handle AIMessage response
        content = response.content if hasattr(response, 'content') else str(response)
        # Clean potential markdown formatting
        content = content.strip()
        if content.startswith("```json"):
            content = content[7:]
        if content.startswith("```"):
            content = content[3:]
        if content.endswith("```"):
            content = content[:-3]
        content = content.strip()
        
        # Robust JSON parsing
        try:
            result = json.loads(content)
        except json.JSONDecodeError:
            start = content.find("{")
            end = content.rfind("}")
            if start != -1 and end != -1 and end > start:
                inner = content[start : end + 1]
                try:
                    result = json.loads(inner)
                except json.JSONDecodeError:
                    result = ast.literal_eval(inner)
            else:
                result = ast.literal_eval(content)

        if not isinstance(result, dict):
            raise ValueError("Parsed analysis is not a JSON object")
        
        # Count critical gaps
        critical_gaps_count = sum(
            1 for key, value in result.items()
            if isinstance(value, dict) and value.get("priority") == "CRITICAL"
        )
        
        result["critical_gaps_count"] = critical_gaps_count
        result["score"] = result.get("overall_score", 0)
        result["framework"] = "ISO/IEC 42001"
        
        return result
        
    except Exception:
        return {
            "framework": "ISO/IEC 42001",
            "score": 0,
            "critical_gaps_count": 0,
            "status": "NOT_EVALUATED",
            "critical_gaps": [],
            "priority_actions": [],
            "strengths": [],
            "raw_response": content if 'content' in locals() else "",
            "compliance_summary": "ISO 42001 analysis could not be generated from the model output. Treat this framework as not yet assessed and review ISO 42001 requirements separately."
        }

File: nextjs-app/src/lib/agents/isoAgent.ts
import { ExtractedData, ISOResult } from '../backend/types';
import { callPerplexity, parseJsonResponse } from '../llm/perplexityClient';

function getISOPrompt(extractedData: ExtractedData): string {
    const docType = extractedData.document_type || 'SYSTEM_SPEC';
    const topicsCovered = extractedData.compliance_topics_covered || [];

    return `
You are an ISO/IEC 42001:2023 (AI Management System) compliance specialist.

**DOCUMENT TYPE: ${docType}**

CRITICAL SCORING GUIDANCE:
- If document_type is "GUIDANCE": Score based on whether it PROVIDES FRAMEWORKS for AI governance, risk management, lifecycle management. Government guidance covering these topics should score WELL.
- If document_type is "SYSTEM_SPEC": Score based on specific organizational implementation.

**DOCUMENT DETAILS:**
- Document type: ${docType}
- Topics covered: ${topicsCovered.length > 0 ? topicsCovered.join(', ') : 'None identified'}
- Use case: ${extractedData.use_case || 'Unknown'}
- Human oversight: ${extractedData.has_human_oversight}

**DOCUMENT TEXT:**
${extractedData.full_text?.slice(0, 25000) || ''}

---

Analyze these ISO 42001 requirement areas:
1. Governance Framework
2. Risk Management
3. Data Quality & Lifecycle
4. Monitoring & Incident Response

Return ONLY valid JSON (no markdown):
{
  "document_type_detected": "${docType}",
  "governance": { "status": "...", "evidence_found": [], "sections_relevant": [], "gap": "", "priority": "..." },
  "risk_management": { ... },
  "data_lifecycle": { ... },
  "monitoring": { ... },
  "overall_score": 0-100,
  "critical_gaps": [],
  "strengths": [],
  "priority_actions": [],
  "compliance_summary": "2-3 sentence summary"
}

IMPORTANT: The "..." tokens in the JSON shape above are ONLY illustrative. In your actual JSON output you MUST fill in complete objects for every field and you MUST NOT output any "...". If information is missing, still provide full objects with empty arrays or explanatory strings.
`;
}

export async function analyzeISOCompliance(extractedData: ExtractedData): Promise<ISOResult> {
    const prompt = getISOPrompt(extractedData);

    try {
        const response = await callPerplexity(prompt, 'sonar-pro');
        const result = parseJsonResponse<ISOResult>(response);

        const overallScore = typeof (result as any).overall_score === 'number' ? (result as any).overall_score : 0;

        const criticalGapsCount = [result.governance, result.risk_management, result.data_lifecycle, result.monitoring].filter(
            (p) => (p as any)?.priority === 'CRITICAL'
        ).length;

        return {
            ...result,
            framework: 'ISO/IEC 42001',
            score: overallScore,
            critical_gaps_count: criticalGapsCount,
            critical_gaps: result.critical_gaps || [],
            priority_actions: result.priority_actions || [],
            strengths: result.strengths || [],
            compliance_summary: result.compliance_summary || 'Analysis completed.',
        };
    } catch (error) {
        console.error('ISO analysis error:', error);
        return {
            framework: 'ISO/IEC 42001',
            score: 0,
            critical_gaps_count: 0,
            critical_gaps: [],
            priority_actions: [],
            strengths: [],
            compliance_summary: 'ISO 42001 analysis could not be completed. Treat this framework as not yet assessed.',
        };
    }
}
